Parse the output to create updated RunDto object.

zpChris opened the pull request:
### List of Changes:

- Create the `RunDto` and `ResultDto` objects to hold the output result and relevant information.
- Create methods to parse the output and store the result in the `RunDto` object.
- Create the `OutputSection` enum to help with parsing.
- Add TODOs to list of tests that cannot be fully implemented.

### Notes:

- The `isOutputCorrect` method just checks the user's output against the test output. Per your notes on #11, our final implementation will be different than this, but I wanted to include some basic placeholder version of this method in this PR.
- Since this PR merged off of `main` (I figured the other branches that are out are not quite finalized, so I thought I'd try to avoid some of those merge conflicts, but I'm open to changing this), the tests are left unimplemented and I added the `DELIMITER` variables in the `DockerService.java` file.
- The `parseCaptureOutput` method is fairly complicated, and part of its inner workings was split off into the helper function `parseTestCaseOutput` in order to break up complexity.
- I use `stringBuilder.setLength(0)` because it is [a fast "clear" method](https://stackoverflow.com/questions/5192512/how-can-i-clear-or-empty-a-stringbuilder).
- I added a field in the `RunDto` object to hold the runtime -- this could be calculated potentially [using `System.currentTimeMillis()`](https://stackoverflow.com/questions/5204051/how-to-calculate-the-running-time-of-my-program).

### Screencast and Images:

There are no changes to the UI and thus no screencasts or images are attached.

zpChris commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Create the result output data structure](https://trello.com/c/ybilCESj/258-create-the-result-output-data-structure)

alankbi commented on src/main/java/com/rocketden/tester/service/DockerService.java:
```diff
@@ -1,53 +1,119 @@
     private static final int TIME_LIMIT = 2;
 
-    public RunDto spawnAndRun(String folder, Language language) {
+    public static final String DELIMITER_TEST_CASE = "###########_TEST_CASE_############";
+    public static final String DELIMITER_SUCCESS = "###########_SUCCESS_############";
+    public static final String DELIMITER_FAILURE = "###########_FAILURE_############";
```
Note: after merging in main, I think you should be able to remove these and use the delimiters in OutputParser

zpChris replied:
Sounds good -- done!

alankbi commented on src/main/java/com/rocketden/tester/service/DockerService.java:
```diff
@@ -1,53 +1,119 @@
         StringBuilder output = new StringBuilder();
+        OutputSection outputSection = OutputSection.START;
+        int numCorrect = 0;
+        int testCount = 0;
```
Is it possible to remove or simplify the way testCount is used? It seems like there's 2 usages of it, 1. `testCases.get(testCount)` and `testCount != testCases.size()`

A potential alternative to both of these is `testCases.get(results.size())` and `results.size() != testCases.size()`, which I feel might help get rid of the need to pass testCount into `parseTestCaseOutput` while also returning testCount from that method.

zpChris replied:
This is a good idea! I removed unnecessary uses of `testCount` and replaced the rest with `results.size()`.

alankbi commented on src/main/java/com/rocketden/tester/service/DockerService.java:
```diff
@@ -1,53 +1,119 @@
-            output.append(s).append("\n");
+            // Update the output section.
+            if (s.equals(DELIMITER_TEST_CASE)) {
+                testCount = parseTestCaseOutput(outputSection,
+                    output.toString(), results, result, testCases, testCount);
```
Combined with the other comment about testCount, is it possible to simplify this maybe by directly passing in `testCases.get(...)` (since you're not using testCases for any other purpose)? Also, instead of passing in both results and result, maybe you could just pass in result (for it to modify) and then just directly add result to results the next line down?

zpChris replied:
The first idea sounds great -- done! ✅ 

The only caveat with the second simplification is when `outputSection = OutputSection.START` (essentially, the first line of the decoding file), we should not be adding anything to the `results` list, as the `result` variable is not yet constructed.

I could add an if statement check for this case to see if `result` should be added, but at that point it might be simpler to pass in the `results` object -- I'll leave this up to you, I'm open to implementing this change or leaving it as-is.

alankbi replied:
No I think what you said makes sense - lets leave it as is then

alankbi commented on src/main/java/com/rocketden/tester/service/DockerService.java:
```diff
@@ -56,4 +122,62 @@ private RunDto captureOutput(Process process) throws IOException, InterruptedExc
+
+            // Update / clear variables that control loop process.
+            testCount++;
+        } else if (outputSection == OutputSection.TEST_CASE) {
+            throw new ApiException(ParserError.MISFORMATTED_OUTPUT);
+        }
+
+        return testCount;
```
nit: Perhaps move this to the top so that it's more clear when an error occurs? And then also same note about the testCount stuff (see other comment)

zpChris replied:
Makes sense -- done! 👍

alankbi commented on src/main/java/com/rocketden/tester/service/DockerService.java:
```diff
@@ -1,53 +1,119 @@
+     * @param problem The current problem associated with this output.
+     * @return The RunDto object produced from the output.
+     */
+    private RunDto parseCaptureOutput(Process process, Problem problem) throws IOException, InterruptedException {
```
Maybe move all this output capturing code to the OutputParser class? I think that can help with single responsibility  stuff (docker service handles docker stuff, output parser handles capturing/judging output)

zpChris replied:
I think that makes sense -- I'll implement this! 👍

zpChris replied:
Ok, I moved it all to the `OutputParser`! I decided not to make it static, as there will probably be a private Gson object in that class, like `InputParser`.

alankbi commented on src/main/java/com/rocketden/tester/dto/RunDto.java:
```diff
@@ -1,11 +1,22 @@
 @Setter
 public class RunDto {
     private boolean status;
-    private String output;
+    private List<ResultDto> results;
+    private String code;
+    private Language language;
+    private Integer numCorrect;
+    private Integer numTestCases;
+    private Long runtime;
+    private LocalDateTime startTime;
```
Could we simplify/remove any number of these fields? For example, not sure what situation status could be false if an ApiException was not already thrown, or also idk if code and language needs to be returned when it'd just be copying them from the request (on main we'll likely also be copying the contents from this response to the Submission object so it's not crucial to have everything in this return if it'll also be on main), or also numTestCases should just be results.size(), and startTime is also calculated on main too (so if we keep it here we'd have to ditch the calculation on main). 

These are just a couple suggestions though - so feel free to take them up on only some/none of them

zpChris replied:
I'll remove the `status` variable -- I want to note, do I still need the following line:
```
process.waitFor(TIME_LIMIT, TimeUnit.SECONDS);
```
This references [this method](https://docs.oracle.com/javase/8/docs/api/java/lang/Process.html#waitFor-long-java.util.concurrent.TimeUnit-), but I'm not sure why (or why not) it is needed, so I wanted to check when removing it.

It'd be good to have a discussion on this, but I thought of adding the `code`, `language`, and `startTime` variables because I imagined that when the main repository gets this object, in the future, it may store this full object in the database. This would then have all the necessary information (besides the User, Problem, and Game, but those may be able to be found by association / relation?), to be placed directly in the database.

If we took those fields out, we could still construct the relevant object, but that would require defining two classes on the main repository, one for accepting this object and one for placing it in the database. Since this seemed repetitive, I thought perhaps all those fields could be placed in one object here.

zpChris replied:
Per our discussion today, I'll remove these unnecessary fields, as on the main repository we'll need both an entity and a DTO class.

alankbi reviewed (changes requested):
Awesome work! I requested a few changes but overall looks like we're 98% of the way to the MVP :) 

Also, a question: this PR is going to be followed up with 1 (or 2?) PRs right? One remaining thing would be the reading in of the serialized outputs and comparing them using Java objects, and the other would be fixing those test cases once that behavior is implemented (as well as making sure the captureParseOutput logic is being tested, as it currently is not here)

alankbi commented on src/main/java/com/rocketden/tester/service/DockerService.java:
```diff
@@ -4,54 +4,30 @@
             Process process = builder.start();
 
-            return captureOutput(process);
+            OutputParser outputParser = new OutputParser();
```
nit: Autowire this class

zpChris replied:
Whoops, I somehow missed that `OutputParser` is a service class -- it's now autowired, thanks for the correction!

alankbi commented on src/main/java/com/rocketden/tester/service/parsers/OutputParser.java:
```diff
@@ -10,9 +23,114 @@
+        String s;
+        while ((s = stdInput.readLine()) != null) {
+            // Update the output section.
+            if (s.equals(DELIMITER_TEST_CASE)) {
+                parseTestCaseOutput(outputSection,
+                    output.toString(), results, result,
+                    testCases.get(results.size()));
+                output.setLength(0);
+                outputSection = OutputSection.TEST_CASE;
```
I must say the VSCode line character limit must be pretty intense haha

zpChris replied:
Haha, yeah it's at 100 but since this is fairly far nested in for loops and if statements, it ends pretty early. Happy to change this convention but yeah, that's why this is split across lines.

alankbi reviewed (approved):
Looks good to me!

zpChris commented:
The code seems to be failing right now when I run the POST request, so I'm investigating now.

zpChris commented:
> The code seems to be failing right now when I run the POST request, so I'm investigating now.

I fixed the issue -- I had to add the last remaining test case output _after_ the while loop ended, as there is not a final "test case delimiter" to end the file. The added code is as follows, immediately after the while loop in `OutputParser.java`:

```
// Add the last remaining test case output, if one exists.
parseTestCaseOutput(outputSection,
    output.toString(), results, result,
    testCases.get(results.size()));
```

zpChris commented:
I've also fixed some additional errors; I've listed them below:
1. After the console for the result is set, I added a line to clear the `output` variable (otherwise it is included in the user's code result): `output.setLength(0);`.
2. The mechanism to set the number of problems correct was not included, so I added this. However, @alankbi, since the primary method `parseCaptureOutput` is essentially at the limit of its cognitive complexity, I instead passed in `numCorrect` into the helper method where it calculated any increments and returned the result, which I can then direct store in the original method. I know you had commented on this practice previously as clunky, so -- thoughts on this?
3. I fixed some of the tests and added an additional test in Python to make sure it runs correctly if the user's code fails.
4. ...

Note: When checking the equality of solutions, we may want to artifically add an `\n` to the correct test case to match the output of the result object.

zpChris commented:
> I do think it's a bit clunky haha - maybe instead if we didn't want to add any more logic to parseCaptureOutput we could simply replace that line with `runDto.setNumCorrect(calculateNumCorrect(results))`? Where calculateNumCorrect is a private helper method that just loops over the results and returns the count. Imo the 0.x% increase in runtime is offset by better readability, but lmk what you think about this/any alternatives as well

Cool, I'll try this out right now -- thanks for responding here. Should I merge after I implement this @alankbi, or should I check with you again?

> This is for current testing purposes right? Cause I know for the follow up PR where we change to use Java equality checks the new lines would prob be trimmed/rendered irrelevant by Gson

That's a good point, that's not really a concern -- thanks!

alankbi commented:
> Cool, I'll try this out right now -- thanks for responding here. Should I merge after I implement this @alankbi, or should I check with you again?

Up to you, although I think it's probably good to just merge since it's not too significant or anything

zpChris commented:
> Up to you, although I think it's probably good to just merge since it's not too significant or anything

Ok, I'll go with this -- thanks! 👍
